### DISCUSSION TITLE
next/router ( Uncaught (in promise) TypeError: handler is not a function ) mitt.js

### DISCUSSION THREAD

**Discussion by himadrinath** [vercel/next.js#38222] [Help]
when Modal (react class component) open in componentDidMount

Router.push('url', 'as', {options})

getting error

```
Uncaught (in promise) TypeError: handler is not a function
    at eval (mitt.js?851d:19:17)
    at Array.map (<anonymous>)
    at Object.emit (mitt.js?851d:18:39)
    at eval (router.js?8684:870:39)
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (router.js?8684:37:28)
    at _next (router.js?8684:55:17)
```

its working until next@12.1.7-canary.39

```
Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 14.17.4
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 12.2.1-canary.1
      eslint-config-next: 11.1.2
      react: 18.2.0
      react-dom: 18.2.0
```

**Comment by icyJoseph** [comment]
> when Modal (react class component) open in componentDidMount

Does this mean you are using  Modal that opens using `#`? I tested this with a JS controlled Modal and it worked fine.

**Reply by himadrinath** [reply]
no not using with #

here are the link [url](https://inmypage.co/himadri)

**Reply by icyJoseph** [reply]
where's the modal? how do you trigger it?

**Reply by himadrinath** [reply]
click on any feed paragraph the modal will appear

**Reply by himadrinath** [reply]
its working and its running on 12.1.7-canary.39 but after update to 12.2.0 getting this error

**Reply by icyJoseph** [reply]
Mmm perhaps you could share some code instead? or a repository with an example?

**Reply by himadrinath** [reply]
ok. working on example repo. thanks

**Reply by himadrinath** [reply]
take look at this example when i open Modal routeChangeError get fire

https://stackblitz.com/edit/nextjs-ow1epj

**Reply by himadrinath** [reply]
@icyJoseph did u looked into this

**Reply by icyJoseph** [reply]
Hi, only briefly, but I'll try to take a deeper look sometime this weekend.

**Reply by himadrinath** [reply]
thanks

**Reply by icyJoseph** [reply]
Alright I think I have an explanation that satisfies me.

This is a Strict Mode issue. Which means it happens only on development.

You can proof this by placing two `console.count`, like this:

```js
  componentDidMount() {
    console.count('PageStatus');

    Router.push(`/?id=${this.props.id}`, null, {
      scroll: false,
    });
  }

  componentWillUnmount() {
    console.count('Unmount PageStatus');
  }
```

When you open the modal you'll see:

```
PageStatus: 1
Unmount PageStatus: 1
PageStatus: 2
```

This is part of the React Strict Mode in version 18, https://reactjs.org/docs/strict-mode.html#ensuring-reusable-state:

> * React mounts the component.
>     * Layout effects are created.
>     * Effect effects are created.
> * React simulates effects being destroyed on a mounted component.
>     * Layout effects are destroyed.
>     * Effects are destroyed.
> * React simulates effects being re-created on a mounted component.
>     * Layout effects are created
>     * Effect setup code runs


So the router abandons the navigation, and somewhat correctly, reports it as an error while changing routes.

If you find this error while developing to be distracting, perhaps you could do:

```js
  componentDidMount() {
    console.count('PageStatus');

    this.raf = window.requestAnimationFrame(() => {
      Router.push(`/?id=${this.props.id}`, null, {
        scroll: false,
      });
    });
  }

  componentWillUnmount() {
    console.count('Unmount PageStatus');
    window.cancelAnimationFrame(this.raf);
  }
```

And that'll silence it. You can try to use an Abort Controller too, but the thing is that the Next.js router event bus will still send the error to all members. 

However, when the error is because there route change was cancelled, you can read the `cancelled` property on the error:

```js
  componentDidMount() {
    Router.events.on('routeChangeError', this.routeChangeError);
  }

  routeChangeError = (err, url, { shallow }) => {
    console.info('%crouter error: ', err);
    console.info('%crouter error: ', url);
    console.log('Cancelled?', err.cancelled);

    if (!err.cancelled) {
      alert('Unexpected error', err);
    }
  };
```

### Summary

- React 18's Strict Mode mounts, unmounts and then mounts again components to test for reusable state
- Next.js' router detect the route cancellation caused by Strict Mode and reports it to its event bus
- In the Router.events, `routeChangeError`, you can read the property `cancelled` on the error, which tells you if the route was simply cancelled
- Alternatively, you can use an animation frame or any other kind of async work, excluding `queuemicrotask`, to delay the execution of the `Router.push` call, till after the `unmount` lifecycle is done
- This is not an issue in production

**Reply by himadrinath** [reply]
@icyJoseph thanks for your brief explanation. 

i also seeing this in production

![router_error_ssr](https://user-images.githubusercontent.com/12247490/177025715-aa861a71-8051-4470-bb90-7ba5fc6a0c76.png)

**Reply by icyJoseph** [reply]
Is it the same error though? What do you see if you log `console.log(err.cancelled)` - perhaps we are looking at two different issues.

**Reply by himadrinath** [reply]
```
console.log(err.cancelled)

undefined
```

i also think so two are different issues
